Competitive_Programming/C++/Binary Trees/lca.md

```diff
@@ -0,0 +1,111 @@
+# Lowest Common Ancestor of Two Nodes
+
+## Problem
+Given a binary tree, with all unique nodes, find the lowest common ancestor (LCA) of two given nodes in the tree.
+
+## Explanation
+Say there are two given nodes of a binary tree, P and Q. The lowest common ancestor of P and Q is defined as the lowest node in the tree that has both P and Q as it's descendants. Note that a node a allowed to be a descendant of itself.
+
+## Example
+
+Input: [3, 5, 1, 6, 2, NULL, 8]
+
+Given: P = 5 and Q = 1
+
+```
+         3
+        / \
+       5   1
+      / \   \
+     6   2   8
+```
+Output: The LCA of 5 and 1 is **3**.
+
+## Approach
+
+The problem can have two major approaches:
+
+* First method: Brute force Tree Traversal. This is done by storing paths from root to node 1 and root to node 2. This methods consumes O(n) time complexity.
+    ### Algortihm
+    1. Create two vectors.
+    2. Store path from root to node 1 in vector1.
+    3. Store path from root to node 2 in vector 2.
+    4. Traverse the vectors till elements are same.
+    5. Return the last same element.
+
+* Second method: Recursive Approach. This method can be adopted to find LCA in a single traversal. The time complexity remains O(n).
+    ### Algorithm
+    1. Traverse tree from root.
+    2. If root == node1 or root == node2, then return root.
+    3. Recursively call the function for left and right subtrees.
+    4. If both the nodes are in left or right subtree exclusively then we will return the left or right subtree respectively.
+    5. If current node is NULL then we have reached the end of tree.
```

Comment:
Place this _below_ approach-1's code and _above_ approach-2's code.
Like:

Description
Code
Description
Code